p2p/protocols/protocol.go

```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
 	}
 }
 
-// Run starts the forever loop that handles incoming messages
-// called within the p2p.Protocol#Run function
-// the handler argument is a function which is called for each message received
+// Run starts the forever loop that handles incoming messages.
+// The handler argument is a function which is called for each message received
 // from the remote peer, a returned error causes the loop to exit
-// resulting in disconnection
+// resulting in disconnection of the protocol
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	wait := p.run(handler)
+	return wait()
+}
+
+// run receives messages from the peer and dispatches async routines to handle the messages
+func (p *Peer) run(handler func(ctx context.Context, msg interface{}) error) func() error {
+	p.mtx.Lock()
+	p.running = true
+	p.mtx.Unlock()
+
+	errc := make(chan error)
+	go func() {
+		for {
+			msg, err := p.readMsg()
+			if err != nil {
+				select {
+				case errc <- err:
+				default:
+				}
+
+				return
 			}
-			return err
+
+			p.mtx.RLock()
+			// if loop has been stopped, we don't dispatch any more async routines and discard (consume) the message
+			if !p.running {
+				_ = msg.Discard()
+				p.mtx.RUnlock()
+				continue
```

Comment:
Again, I did it this way to keep the same logic as before. The loop was still running without executing the active handlers. I guess the reason for this is because the "graceful shutdown" was waiting for async handlers to finish before returning the loop.